internal/flow/import_git_test.go

```diff
@@ -84,27 +142,178 @@ testImport.add "cc" {
+	testRepo := t.TempDir()
+
+	runGit(t, testRepo, "init", testRepo)
+	math := filepath.Join(testRepo, "math.river")
+	err := os.WriteFile(math, []byte(contents), 0666)
+	require.NoError(t, err)
+
+	runGit(t, testRepo, "add", ".")
+
+	runGit(t, testRepo, "commit", "-m \"test\"")
+
+	getHead := exec.Command("git", "rev-parse", "HEAD")
+	var stdBuffer bytes.Buffer
+	getHead.Dir = testRepo
+	getHead.Stdout = bufio.NewWriter(&stdBuffer)
+	err = getHead.Run()
+	require.NoError(t, err)
+	hash := stdBuffer.String()
+	hash = strings.TrimSpace(hash)
+
+	main := `
+import.git "testImport" {
+	repository = "` + testRepo + `"
+  	path = "math.river"
+    pull_frequency = "10s"
+    revision = "` + hash + `"
+}
+
+testImport.add "cc" {
+	a = 1
+    b = 1
+}
+`
+
+	// After this update the sum should still be 2 and not 3 since it is pinned to the initial hash.
+	err = os.WriteFile(math, []byte(contentsMore), 0666)
+	require.NoError(t, err)
+
+	runGit(t, testRepo, "add", ".")
+
+	runGit(t, testRepo, "commit", "-m \"test2\"")
+
+	defer verifyNoGoroutineLeaks(t)
+	ctrl, f := setup(t, main)
+	err = ctrl.LoadSource(f, nil)
+	require.NoError(t, err)
+	ctx, cancel := context.WithCancel(context.Background())
+
+	var wg sync.WaitGroup
+	defer func() {
+		cancel()
+		wg.Wait()
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		ctrl.Run(ctx)
+	}()
+
+	// Check for final condition.
+	require.Eventually(t, func() bool {
+		export := getExport[map[string]interface{}](t, ctrl, "", "testImport.add.cc")
+		return export["sum"] == 2
+	}, 20*time.Second, 1*time.Millisecond)
+}
+
+func TestPullUpdatingFromTag(t *testing.T) {
+	testRepo := t.TempDir()
+
+	runGit(t, testRepo, "init", testRepo)
+
+	math := filepath.Join(testRepo, "math.river")
+	err := os.WriteFile(math, []byte(contents), 0666)
+	require.NoError(t, err)
+
+	runGit(t, testRepo, "add", ".")
+
+	runGit(t, testRepo, "commit", "-m \"test\"")
+
+	runGit(t, testRepo, "tag", "-a", "tagtest", "-m", "testtag")
+
+	main := `
+import.git "testImport" {
+	repository = "` + testRepo + `"
+  	path = "math.river"
+    pull_frequency = "10s"
+    revision = "tagtest"
+}
+
+testImport.add "cc" {
+	a = 1
+    b = 1
+}
+`
+
+	// After this update the sum should still be 2 and not 3 since it is pinned to the tag.
+	err = os.WriteFile(math, []byte(contentsMore), 0666)
+	require.NoError(t, err)
+
+	runGit(t, testRepo, "add", ".")
+
+	runGit(t, testRepo, "commit", "-m \"test2\"")
+
+	defer verifyNoGoroutineLeaks(t)
+
+	ctrl, f := setup(t, main)
+	err = ctrl.LoadSource(f, nil)
+	require.NoError(t, err)
+	ctx, cancel := context.WithCancel(context.Background())
+
+	var wg sync.WaitGroup
+	defer func() {
+		cancel()
+		wg.Wait()
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		ctrl.Run(ctx)
+	}()
+
+	// Check for final condition.
+	require.Eventually(t, func() bool {
+		export := getExport[map[string]interface{}](t, ctrl, "", "testImport.add.cc")
+		return export["sum"] == 2
+	}, 20*time.Second, 1*time.Millisecond)
```

Comment:
in the file, you use in several places a pull-frequency of 10s and a timeout of 20 seconds. Could both durations be reduced?

Reply:
Good call